client/asset/xmr/toolsdl/download.go

```diff
@@ -0,0 +1,613 @@
+
+	// chmod +x
+	err = d.setExecutable(toolsDir, isWin)
+	if err != nil {
+		d.cleanAll()
+		return toolsDir, err
+	}
+	d.log.Trace("Set binaries as executable")
+
+	// all tools files downloaded, checked, archive selected and checked sha256;
+	// files created and made executable
+	d.cleanup()
+	d.log.Trace("Cleaned up all un-needed files")
+	d.log.Debugf("Tools Dir: %s", toolsDir)
+	return toolsDir, nil
+}
+
+func downloadHashesFile(ctx context.Context) (string, error) {
+	url := HashesLink
+	toolsPath, err := getToolsBasePath()
+	if err != nil {
+		return "", err
+	}
+	err = os.MkdirAll(toolsPath, 0700)
+	if err != nil {
+		return "", err
+	}
+	hashesFilePath := filepath.Join(toolsPath, MoneroHashesFilename)
+	hashesFile, err := os.Create(hashesFilePath)
+	if err != nil {
+		return "", err
+	}
+	defer hashesFile.Close()
+
+	// HTTP GET - TODO(dl) certs
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", err
+	}
+
+	_, err = io.Copy(hashesFile, resp.Body)
+	if err != nil {
+		return "", err
+	}
+	return hashesFilePath, nil
+}
+
+func (d *Download) getHashedZips(hashesFilePath string) error {
+	d.z = make(hashedZips, 0)
+	hf, err := os.Open(hashesFilePath)
+	if err != nil {
+		return err
+	}
+	defer hf.Close()
+
+	scanner := bufio.NewScanner(hf)
+
+	gatheringHashes := false
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.Contains(line, CliTrigger) {
+			gatheringHashes = true
+			continue
+		} else if gatheringHashes {
+			if line == HashZipTerminator {
+				break
+			}
+			tkns := strings.Split(line, TwoSpaces)
+			if len(tkns) != 2 {
+				return fmt.Errorf("got %d tokens, expected 2", len(tkns))
+			}
+			d.z = append(d.z, hashedZip{
+				hash: tkns[0],
+				zip:  tkns[1],
+			})
+		}
+	}
+
+	return scanner.Err()
+}
+
+func (d *Download) checkHashedZips() error {
+	for i := range d.z {
+		d.z[i].valid = true
+		tkns := strings.Split(d.z[i].zip, Dash)
+		if len(tkns) > 4 && len(tkns) < 3 {
+			return fmt.Errorf("checking %s got %d tokens from %s, expected 3 or 4", d.z[i].zip, len(tkns), Dash)
+		}
+		if tkns[0] != "monero" { // all files start with monero-*
+			d.z[i].valid = false
+			continue
+		}
+		if tkns[1] == "source" || tkns[1] == "android" { // unused os's
+			d.z[i].valid = false
+			continue
+		}
+		if tkns[2] == "x86" { // no more intel 32bit arch's; sort out the arm ones then 99% done!
+			d.z[i].valid = false
+			continue
+		}
+
+		d.z[i].os = tkns[1]
+		d.z[i].arch = tkns[2]
+
+		verExtTkns := strings.Split(tkns[3], Dot)
+		if len(verExtTkns) < 5 || len(verExtTkns) > 6 {
+			return fmt.Errorf("checking %s got %d version/extension tokens, expected 5 or 6", d.z[i].zip, len(verExtTkns))
+		}
+
+		verTkns := strings.SplitN(tkns[3], Dot, 5)
+		if len(verTkns) != 5 {
+			return fmt.Errorf("checking %s got %d version+ tokens, expected 5", d.z[i].zip, len(verTkns))
+		}
+		d.z[i].ext = verTkns[4]
```

Comment:
Didnt you just split this and error if not 5 or 6? Looks like six if ends in .bz2